### DISCUSSION TITLE
Sign the un-preprocessed `git archive` .tar.gz

### DISCUSSION THREAD

**Discussion by kpcyrd** [curl/curl#13520] [General]
### I did this

(First of all, sorry for being so late, this is a followup on "Reproducible tarballs" from [this blogpost](https://daniel.haxx.se/blog/2024/04/10/verified-curl/))

The steps needed to "[reproduce the source code tarballs](https://github.com/curl/curl/pull/13250)" is quite involved and I consider autotools pre-processing part of the regular build process that should happen on the Linux distro's build servers (so it's covered by reproducible builds infrastructure that e.g. [Arch Linux operates](https://reproducible.archlinux.org/)).

The tarball Github provides is already reproducible, you can download a `git archive` snapshot of the git tree-ish from here:

```
https://github.com/curl/curl/archive/refs/tags/curl-8_7_1.tar.gz
```

I couldn't find a corresponding `curl-8_7_1.tar.gz.asc`, but I think there should be one.

You don't need to sign the tarball downloaded from Github, you can generate it locally:

```
git -c tar.tar.gz.command="gzip -cn" archive --prefix="curl-curl-8_7_1/" -o curl-8_7_1.tar.gz curl-8_7_1
```

But since the files are supposed to 100% match, the signature would be valid for the Github generated tarball too.

If the prefix and compression bother you, I'd recommend to sign this instead (note `.tar` instead of `.tar.gz`):
```
git archive -o curl-8_7_1.tar curl-8_7_1
```

Since `git archive` is deterministic, the generated file is expected to have a sh256 of `d889aa963bb0d5f55801d122bda82f841e7807242195c5ceed2d30fdccfd2595` (given `curl-8_7_1` points to de7b3e89218467159a7af72d58cea8425946e97d). This hash is also how Arch Linux references the source code of curl:

https://whatsrc.org/artifact/sha256:d889aa963bb0d5f55801d122bda82f841e7807242195c5ceed2d30fdccfd2595

While most of the other distros use the autotools pre-processed (but signed) tarball:

https://whatsrc.org/artifact/sha256:99032a437fbcb38a37f35f23712bbeb6489be112450e02443e16786d3d745b31

Thanks for your attention!

### I expected the following

There should be a signature on the pristine source code snapshot taken with `git archive`.

### curl/libcurl version

8.7.1

### operating system

Arch Linux

**Comment by bagder** [comment]
> I consider autotools pre-processing part of the regular build process that should happen on the Linux distro's build servers

There are also many other people who build curl from source. They don't need to have the autotools installed.

## on GitHub tarballs

Given GitHub's own comments on [reproducibility of their tarballs](https://github.blog/2023-02-21-update-on-the-future-stability-of-source-code-archives-and-hashes/), they seem less certain: *"If you rely on stable archives for security (ensuring you don’t accidentally trigger a tarbomb, for example), we recommend you switch to release assets instead of using source downloads"*

I hesitate starting to sign GitHub tarballs because I fear that it risks increasing the confusion about what a curl release is. A release is a release tarball. They come in four different compression formats. They are signed.

To generate a release-build directly from a git tag, you can't just check out the code and build it since it will not match what the corresponding tarball. I believe we don't even document that procedure since we don't encourage taking that route. The version and timestamp need to be set in `include/curl/curlver.h` and we used to produce/update a few other files as well.

**Reply by kpcyrd** [reply]
> There are also many other people who build curl from source. They don't need to have the autotools installed.

I'm not asking you to stop providing tarballs with pre-generated configure scripts, I'm instead asking you to sign a deterministic VCS snapshot **in addition**. :)

You're currently making people choose between

- having a signed tarball
- not having to code review 49k lines of auto-generated `./configure` shell script
- not having to deal with git to authenticated the code through signed git tag 

Out of those your users currently need to "pick two" but can't have all three.

> Given GitHub's own comments on [reproducibility of their tarballs](https://github.blog/2023-02-21-update-on-the-future-stability-of-source-code-archives-and-hashes/), they seem less certain

I remember this incident, it was about changes in git's default `gzip` compression code, this is why I suggested to alternatively consider signing the uncompressed `.tar`. The actual `.tar` output did not change back then, and it's less likely to do so in the future. There was even a statement from the product manager for Git at Github back then:

https://news.ycombinator.com/item?id=34587306

You're currently working around this in "reproducible autotools tarballs" by specifying `debian:bookworm-slim@sha256:993f5593466f84c9200e3e877ab5902dfc0e4a792f291c25c365dbe89833411f` in your `Dockerfile` for release automation, which implicitly references `GNU gzip 1.12`, but at that point `git archive` would also be 100% guaranteed to be reproducible, even with compression enabled.

> I hesitate starting to sign GitHub tarballs because I fear that it risks increasing the confusion about what a curl release is. 

There is already confusion about "what is a source code release" in any autotools project, because the concept of "we need to pre-process the source code" is not a thing in most other language ecosystems, any kind of processing is considered part of "the build process".

People commonly expect "the source code" to be "the thing I get from git", this includes [oss-fuzz](https://github.com/google/oss-fuzz/blob/o3/projects/curl/Dockerfile) for example.

> I believe we don't even document that procedure since we don't encourage taking that route.

This kind of "accidental gap" between "the VCS" and "the reproducible builds project" is explicitly what I'm trying to stop from occurring, by encouraging the major distributions to run the build in it's entirety, autotools included.

Arch Linux handles it like this, which I think is reasonable (also note both "the version" and `$SOURCE_DATE_EPOCH` are typically provided by the build servers of the major Linux distributions):

```sh
# no '-DEV' in version, release date from tagged commit...
sed -i \
  -e "/\WLIBCURL_VERSION\W/c #define LIBCURL_VERSION \"${pkgver}\"" \
  -e "/\WLIBCURL_TIMESTAMP\W/c #define LIBCURL_TIMESTAMP \"$(git log -1 --format=%cs "curl-${pkgver//./_}")\"" \
  include/curl/curlver.h

autoreconf -fi 
```

Right now, if one doesn't want autotools pre-processing, one would need to sacrifice "upstream's blessing" by using a .tar that no signature exists for. Debian doesn't natively support the concept of signed git tags and would instead prefer a signed `.tar`.
